### PR TITLE
Design Panel: Add contrast warning

### LINKED ISSUE #7230: Editor: Add a contrast check to the "style" panel
## Feature Description

If a user were to pick a style that would make the text hard to read, announce `This color combination may be hard for people to read`.

The `useLiveRegion` hook can be used to announce the message to the user. However, we will need to implement some sort of contrast check in the panel.

## Alternatives Considered

n/a

## Additional Context

Split off from #6059 due to scope creep

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by BrookeGraham**
## Context

We want to notify users when they choose a text color and background color combination that doesn't meet accessibility. 

## Summary

When a user chooses a font color or background color that doesn't have sufficient contrast, we want to show some warning text where they pick the colors. Now inside the Design Panel -> Text we are showing a warning message. We also are making use of our `useLiveRegion` hook to announce the message as it appears. 
I also added the warning to the `background` panel. 

## Relevant Technical Choices

Created a `Warning` component that can be reused with custom messages as we find other areas where we want to show feedback sooner than in the checklist. 
Added a constant for the warning messages under `ACCESSIBILITY_COPY`
Moved `getFailingPages` into the `checklist/checks/pageBackgroundLowTextContrast/check.js` file so that we can use it in both in the checklist check, and here in the panel. Also, renamed it to `getPagesWithFailedContrast`
\`pageBackgroundTextLowContrast(page)` (which is called inside `getPagesWithFailedContrast` ) now returns the offending elements' id. This way we know exactly which elements should be flagged.

The reason I am checking the whole page for contrast issues is because we could have more than one text selection. If either of them doesn't have enough contrast, we should show the message. Plus, we already need to grab all elements from the page to determine which ones are considered as part of the background. 

## To-do
Some background images may flag for low contrast. I am not sure how we want to accommodate for this. 

## User-facing changes
Screen Shot 2021-12-31 at 12 50 11 PM
Screen Shot 2021-12-31 at 12 50 22 PM
Screen Shot 2021-12-31 at 12 50 42 PM
Screen Shot 2021-12-31 at 12 53 21 PM

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. In editor add some text. 
2. Change the font color to match the background. 
3. See the warning appear. 
4. Change it back and the warning should disappear. 
5. Check that the new message is announced. 
6. Turn on voice over, or other reader. 
7. Follow above steps, you should hear the message announced. 
8. In addition, the warning should appear on the background panel. 
9. Click the background on the canvas. The background panel should open. 
10. Change the color to low contrast and you should see the warning message. 
11. The check should still work as normal in the Pre-publish Checklist as well. 

### Does this PR have a security-related impact?
no

### Does this PR change what data or activity we track or use?
no

### Does this PR have a legal-related impact?
no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7230

### CODE REVIEW

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrookeGraham) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: Added the warning message here as well cause I figured why not 🤷🏻‍♀️

File: packages/story-editor/src/components/panels/design/warning/constants.js
Comment (BrookeGraham) [inline packages/story-editor/src/components/panels/design/warning/constants.js]: I added the bit on `font size` because that can affect whether or not the element passes accessibility.

File: packages/design-system/src/icons/warning_low.svg
Comment (BrittanyIRL) [inline packages/design-system/src/icons/warning_low.svg]: icon exists in design system already as `ExclamationTriangle`
Screen Shot 2022-01-03 at 10 53 48 AM

You can run `npm run storybook` and go to design system -> icons to see what we've got.

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js]: nit: 
```suggestion
 * Returns all failing pages as well as the offending elements' id(s).
```

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js]: nit: could this be renamed as like `getPagesWithFailedContrast` or something that's clearer since it's now getting used outside of `/pageBackgroundLowTextContrast` which previously provided more context.

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: thoughts on `import { default as Warning, COLOR_COMBINATION } from '../warning';` ?

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: currentPage will always be index 0?

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: This should work the same, i believe. 

```suggestion
      setFailedContrast(Boolean(result));
```

File: packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js:38]: thanks for writing more tests for this! 

Would you mind adding a second check for any aXe violations on the warning panel here? 
It'll look something like: 

````
const contrastWarning = fixture.editor.inspector.designPanel.queryByTitle('Low Warning');
await expectAsync(contrastWarning).toHaveNoViolations();
```

File: packages/story-editor/src/components/panels/design/textStyle/textStyle.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/textStyle/textStyle.js]: You added `getFailingPages` to exports on `pageBackgroundLowTextContrast` so you could probably import from there too.
```suggestion
import { getFailingPages } from '../../../checklist/checks/pageBackgroundLowTextContrast';
```

File: packages/story-editor/src/components/panels/design/textStyle/textStyle.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/textStyle/textStyle.js]: Samesies here as earlier: 
`import { default as Warning, COLOR_COMBINATION } from '../warning';`

File: packages/story-editor/src/components/panels/design/warning/constants.js
Comment (BrittanyIRL) [reply packages/story-editor/src/components/panels/design/warning/constants.js]: 🤔 I'm wondering if this new set of constants belongs closer to the contrast check that it's giving readable user context too - we could put it in `checklist/constants` under `ACCESSIBILITY_COPY.lowContrast` as `panel` or something else that indicates that it's getting used as a corresponding notification of the check in a panel.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/warning/warning.js]: Since you've got `Message` as `Text` these shouldn't be necessary

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/warning/warning.js]: If this is a new color, can you add it to the design system? Or if it's similar to another color we've got in there, let's use that instead to keep the color palette tight(ish).

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/warning/warning.js:44]: if you swap to the `ExclamationTriangle` default width is 32px which should work the same.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/warning/warning.js]: 🤔  should this be stored with the copy that is `message` for a single source of truth on copy?

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
Stoked you got to this last week! Love the added notes and how you extended the existing functionality. Couple small things.

File: packages/design-system/src/icons/warning_low.svg
Comment (BrookeGraham) [reply packages/design-system/src/icons/warning_low.svg]: Ah nice. Thanks.

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: Yea. If we send in the `currentPage`, it will only check that single page and thus the returned array will have a max length of one. If the `currentPage`  passes, it returns an empty array.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: I couldn't find any thing of the sort, happy to add.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: Good looking out. I may have added this before I was using `Text` I'll double check it isn't needed.

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js
Comment (samwhale) [inline packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js]: 💭 I think rather than updating to add a new argument, we could pass in the current page in its own array using the same function as before:

```js
const failedPages = getFailingPages([currentPage], pageSize);
const didCurrentPageFail = failedPages[0];
```

This reduces the surface area we'd need to test and keeps the function a little more simple. Thoughts?

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: 🤔 I think we don't need this effect to run when `pages` changes. However, if we use the `getFailingPages` function like this:

```getFailingPages([currentPage], pageSize)```

then this effect won't be run when `pages` changes

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (samwhale) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: I think it makes sense to put here. Maybe we should slightly alter the copy to make it specific to background color vs text color. May be a good question for @agingoldseco though:

```
Background: Some text on the page may be hard for people to read against this background. Update the background color to increase the contrast.

Text: This text may be hard for people to read. Update the color or the font size to increase the contrast.
```

File: packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js]: very nice  👍

File: packages/story-editor/src/components/panels/design/textStyle/textStyle.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/textStyle/textStyle.js]: 💭 It would be ideal if this effect only ran when `currentPage` changed (like before).

I think we could derive whether the selected text element has a contrast warning rather than setting it here.

```js
  useEffect(() => {
    const selectedElementIds = selectedElements.map(
      (selectedElement) => selectedElement.id
    );

    getFailingPages(pages, pageSize, currentPage).then((failedPages) => {
      // getFailingPages returns an array of pages, since we only care
      // about currentPage, we can grab the single page result.
      const result = failedPages[0]?.result;
      setFailedPages(result); // set failed ids in state
    });
  }, [currentPage, pageSize, pages, selectedElements]);
...

const showContrastWarning = failedPages?.some((elementId) => selectedElementIds.includes(elementId));

return (
    ...
    {showContrastWarning && <Warning message={COLOR_COMBINATION} />}
    ...
)
```

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/warning/warning.js:56]: nice!

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/warning/warning.js]: 💭 I feel like this font size is really big compared to the rest of the panel font sizes. Should we make it a little smaller? @agingoldseco

**Review by samwhale — COMMENTED** [review]
This works great. I left a few ideas about how we may not need to update `getFailingPages` but rather change the way we use it.

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js]: That makes total sense, I don't know why I made it harder than it should be.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: Agreed

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: Ooops this was my bad. Updated to match figma ie 14px.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: I added `warning: {yellow: #f4b844}` Don't know if that is how we want to add this. We didn't have anything similar so I am unsure. Do I need to add something to Figma as well?

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/check.js]: This was so obvious I seriously don't know why I did the other thing 🤦🏻‍♀️ Thanks so much @samwhale

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: Happy to update copy to whatever y'all think is best. I am not great at copy so I have no preference.

**Comment by BrookeGraham** [comment]
@agingoldseco What do you think about the copy on the background panel vs the text panel? Also how do you feel about the copy including a bit about font size? ie text panel: `This color combination or font size may be hard for people to read.` and background panel: `This background color may make text hard to read.` or something.

File: packages/story-editor/src/components/panels/design/warning/warning.js
Comment (BrittanyIRL) [reply packages/story-editor/src/components/panels/design/warning/warning.js]: you don't need to add to figma - the theme colors populate in storybook: https://google.github.io/web-stories-wp/storybook/?path=/story/designsystem-colors--default 

I think that this yellow goes with the `status` section of colors, keeping colors organized by purpose can allow for easier updates. If we call it yellow and then maybe someday (just an example) the designs change the warning color to be an orange, now that reads wrong because the yellow is actually orange and we'd have to update the instances of that instead of just the theme color.

File: packages/design-system/src/theme/colors.js
Comment (BrittanyIRL) [inline packages/design-system/src/theme/colors.js]: let's move this up to `status`  (line 94) - `warning: '#f4b844'`

File: packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/textStyle/karma/contrast.karma.js]: 😻

**Review by BrittanyIRL — COMMENTED** [review]
Que bonita. Let's wait for design confirmation on copy. Otherwise, just the theme warning color getting put into `status` instead of a new object.

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js
Comment (BrookeGraham) [inline packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js]: Oops bad rebase. Fixing.

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js
Comment (BrookeGraham) [inline packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js]: I had to fix a bad rebase, so while I was in here I updated the `waitFor`s too be a little more robust and actually work the way we want them too.

**Comment by agingoldseco** [comment]
>What do you think about the copy on the background panel vs the text panel? Also how do you feel about the copy including a bit about font size? ie text panel: `This color combination or font size may be hard for people to read.` and background panel: `This background color may make text hard to read.` or something.

I like the bit about font size and separating out the warning types! But would a warning, more often than not, show up in both the background and text panels at once? 

I think this copy works, would just add the word "some" to soften it a bit? `This color combination or font size may be hard for some people to read.` and `This background color may make some text hard to read.` —does that make sense?

**Comment by BrookeGraham** [comment]
> But would a warning, more often than not, show up in both the background and text panels at once?

Yep, @agingoldseco  it would be displayed in both places, but those two panels are not ever open at the same time. Depending on what element is select, the corresponding panel would show. ie. if you have the background selected, you wouldn't see the text panel info / warning.

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js:270]: Thanks for all the updates! Was giving it one more sweep and seeing this:

I added some text that's the same color as an image (black). The contrast check shows correctly on the text but the `page background` also shows the warning even though the text is fine against the background.

May be more of a UX question but I think we should probably show the warning on the element that the text is against rather than the background in this case: (text doesn't show because it's the same color as the black part of the image)

image

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js:270]: I see what you are saying. In this case the "background" doesn't really matter because the text is on top of the image. Making the image the _actual_ background. Therefore the warning shouldn't appear on the background panel. 

There are going to be a ton of edge cases like this. I am not sure how we are going to handle those. 🤔

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js:270]: I'll look into only showing the message here if the failure is cause by background color as opposed to images. Also, the message being here wasn't a requirement so happy to remove it if we decide it's too confusing.

Comment (samwhale) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js:270]: Totally - I don't want any scope creep here.

I think either of these would be good:
1. Always the contrast warning from the page background panel
2. Only removing the contrast warning from the page background panel when the text contrast problem isn't from the background

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js:270]: Such a great catch @samwhale I have updated to only show on the background panel if the contrast problem stems from the actual background element.

**Review by samwhale — APPROVED** [review]
Thanks for making those fixes, works great!

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (maxyinger) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: Hmph thinking about it now, I'm wondering if the background is a good candidate for this check or not. The primary scenario I'm thinking about is:
- you have 2 text elements, one has high enough contrast with the background, the other does not
- you select the background and see that it has a warning that says it doesn't have enough contrast, and the user should change the color
- the user changes the color to better contrast with the original conflicting text, but now the background is conflicting with the other text element (that didn't throw an error before)
- to the user, they changed the color, but the warning just never went away

Comment (maxyinger) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: ^^ not sure if this is a likely scenario, but curious if it's one we should consider.

Comment (maxyinger) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: anyway, I don't want to throw a wrench in this PR which has already gotten design approval. I think if we find ourselves running into the scenario above more frequently, we can always do a little followup PR. Just some behavior to keep our eyes out for in the future

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js
Comment (maxyinger) [inline packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js:165]: if this expect was only ever in here for the `waitFor` statement, I wonder if we even need it anymore?

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js
Comment (maxyinger) [inline packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js]: Ahh I think this may be a bug.

Lets run through the scenario:
- a story with one page that has a white background, a black square, and some black text, with the black text positioned over the background
- you move the text to be on top of the black square
- this triggers a failing page
- you open the help center and see this check
- with the help center open, you move the text to be positioned back over the background
- this check should go away, but doesn't because it only sets state if `failedPages.length`

Comment (maxyinger) [reply packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js]: Also small (and probably very rare) race case to consider is if a user is able to update the story multiple times before this check completes.

We may want some sort of stale flag to ensure that we're always deriving failed pages off of the freshest version of `storyPages`. Something like:
```js
useEffect(() => {
  // can name this whatever you want
  let isInputStale = false;
  getPagesWithFailedContrast(storyPages, pageSize)
      .then((failures) => {
        if (isInputStale) {
          return;
        }
        const failedPages = failures.map(({ page }) => page);
        setFailingPages(failedPages);
      })
      .catch(() => {
        // maybe add a TODO comment in here or something :shrug:
      });
     
  return () => {
    isInputStale = true;
  }
}, [storyPages, pageSize]);
```

**Review by maxyinger — CHANGES_REQUESTED** [review]
Other than the small bug I pointed out, this looks great and works as expected. Happy to slap an approval on this after that's fixed.

**Small note on the implementation chosen here:**
The only thing I'd watch out for is that the contrast check was the most expensive check in the PPC. It was a big part of why we had to refactor the PPC.

The solution proposed here, basically doubles the expensive computation by performing it both in the PPC and in the design panel. We should keep our eyes out for this, and we may need to figure out a solution in the future that performs this expensive calculation once and stores it some place we can reference it vs performing this calculation everywhere we need it referenced.

Anyway, we still haven't performed an audit of the design panels yet. If this does make a noticeable impact we can always update the approach then

File: packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/checklist/checks/pageBackgroundLowTextContrast/component.js]: Great catch. Not sure how I overlooked that.

**Review by maxyinger — APPROVED** [review]

File: packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/pageBackground/pageBackground.js]: Yeah I had thought about that too. My thinking was that regardless of which text element is causing the error, there is still an error and they should find a background that is suitable for all text if they don't want to see the message. The copy here was updated to `This background color may make some text hard to read.` Which I thinks helps convey that _some_  not necessarily all text may be hard to read.

File: packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js
Comment (BrookeGraham) [reply packages/story-editor/src/components/colorPicker/karma/colorPicker.karma.js:165]: Above we are clicking the `addText` button. Since these test can run kinda fast in ci, the node hasn't always been mounted to the dom by the time we need to interact with it. So we added the `waitFor` to force karma to repoll until the node appears on the dom. At least that is my understanding. Otherwise we end up with a flakey test that doesn't always find the node it is looking for.

**Comment by BrookeGraham** [comment]
>The solution proposed here, basically doubles the expensive computation by performing it both in the PPC and in the design panel. We should keep our eyes out for this, and we may need to figure out a solution in the future that performs this expensive calculation once and stores it some place we can reference it vs performing this calculation everywhere we need it referenced.

Totally agree! It is a lot. 

>The only thing I'd watch out for is that the contrast check was the most expensive check in the PPC. It was a big part of why we had to refactor the PPC.

While I was testing, it appears that the checklist mounts and calculates this function regardless of whether or not the checklist 'panel' is open. It then reruns the `getFailingPages` (well, the whole `AccessibilityChecks`) check every time it is opened and closed. 

@littlemilkstudio ☝️

**Comment by BrittanyIRL** [comment]
this test failure isn't because of this pr - looks like the default templates got updated in two PRs and one didn't get tests re run after the other so the failure didn't show up: https://github.com/google/web-stories-wp/pull/10078 and https://github.com/google/web-stories-wp/pull/10217   

(new issue to fix it: https://github.com/google/web-stories-wp/issues/10263)

https://github.com/google/web-stories-wp/runs/4871081344?check_suite_focus=true#step:6:386
Screen Shot 2022-01-19 at 10 18 52 AM

**Review by BrittanyIRL — APPROVED** [review]

**Pull request merged**
